File: wear/src/main/java/com/codegy/aerlink/utils/PacketProcessor.java
package com.codegy.aerlink.utils;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;

import java.io.ByteArrayOutputStream;
import java.math.BigInteger;

/**
 * Created by Guiye on 20/5/15.
 */
public class PacketProcessor {

    private ByteArrayOutputStream processingData;
    private int action;
    private int status;
    // The number of bytes left to process
    private int bytesLeftToProcess;

    public PacketProcessor(byte[] packet) {
        if (packet.length >= 2) {
            action = packet[0];
            status = packet[1];

            if (status != 0x00 && packet.length > 4) {
                byte[] byteLength = {packet[2], packet[3]};
                BigInteger length = new BigInteger(byteLength);
                bytesLeftToProcess = length.intValue();

                Log.d("PacketProcessor", "DATA length " + length);

                processingData = new ByteArrayOutputStream();
                processingData.write(packet, 4, packet.length - 4);

                bytesLeftToProcess -= (packet.length - 4);
            }
        }
    }

    public void process(byte[] packet) {
        processingData.write(packet, 0, packet.length);
        bytesLeftToProcess -= packet.length;
    }

    public int getAction() {
        return action;
    }

    public int getStatus() {
        return status;
    }

    public boolean isFinished() {
        return bytesLeftToProcess <= 0;
    }

    public byte[] getValue() {
        return processingData.toByteArray();
    }

    public String getStringValue() {
        String value = null;
        try {
            value = new String(processingData.toByteArray(), "UTF-8");
        }
        catch (Exception e) {
            e.printStackTrace();
        }

        return value;
    }

    public Bitmap getBitmapValue() {
        Bitmap value = null;
        try {
            value = BitmapFactory.decodeByteArray(processingData.toByteArray(), 0, processingData.toByteArray().length);
        }
        catch (Exception e) {
            e.printStackTrace();
        }

        return value;
    }

}

File: wear/src/main/java/com/codegy/aerlink/utils/PacketProcessor.kt
package com.codegy.aerlink.utils

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import java.io.ByteArrayOutputStream
import java.math.BigInteger
import java.nio.charset.StandardCharsets

class PacketProcessor(packet: ByteArray) {
    private val processingData: ByteArrayOutputStream = ByteArrayOutputStream()
    var action: Int = 0
        private set
    var status: Int = 0
        private set
    private var bytesLeftToProcess: Int = 0
    val isFinished: Boolean
        get() = bytesLeftToProcess <= 0
    val value: ByteArray
        get() = processingData.toByteArray()
    val stringValue: String?
        get() {
            return try {
                String(processingData.toByteArray(), StandardCharsets.UTF_8)
            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    val bitmapValue: Bitmap?
        get() {
            return try {
                BitmapFactory.decodeByteArray(processingData.toByteArray(), 0, processingData.toByteArray().size)
            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }

    init {
        if (packet.size >= 2) {
            action = packet[0].toInt()
            status = packet[1].toInt()

            if (status != 0x00 && packet.size > 4) {
                val byteLength = byteArrayOf(packet[2], packet[3])
                val length = BigInteger(byteLength)
                bytesLeftToProcess = length.toInt()

                Log.d(LOG_TAG, "DATA length $length")
                process(packet, 4)
            }
        }
    }

    fun process(packet: ByteArray, offset: Int = 0) {
        processingData.write(packet, offset, packet.size - offset)
        bytesLeftToProcess -= packet.size - offset
    }

    companion object {
        private val LOG_TAG = PacketProcessor::class.java.simpleName
    }
}
